OpenROAD flow variables document (#1479)
The-OpenROAD-Project/OpenROAD: docs/user/FlowVariables.md

|--------------------------------------|--------------|--------------|---------------|-----------|
| Library Setup                        |              |              |               |           |
| `PROCESS`                            | =            | =            | =             | =         |
| `TECH_LEF`                           | =            | =            | =             | =         |
| `SC_LEF`                             | =            | =            | =             | =         |
| `LIB_FILES`                          | =            | =            | =             | =         |
| `GDS_FILES`                          | =            | =            | =             | =         |
| `DONT_USE_CELLS`                     | =            | =            | =             | =         |
| Synthesis                            |              |              |               |           |
| `LATCH_MAP_FILE`                     | =            | =            | =             | =         |
| `CLKGATE_MAP_FILE`                   | =            | =            | =             | =         |
| `ADDER_MAP_FILE`                     | ?=           | ?=           | ?=            | ?=        |

> | `BLACKBOX_MAP_TCL`                   | N/A          | N/A          | N/A           | =         |

Reviewer: `BLACKBOX_MAP_TCL` is deprecated. There shouldn't be any platforms using this.

| `TIEHI_CELL_AND_PORT`    | Tie high cells used in Yosys synthesis to replace a logical 1 in the Netlist.              |
| `TIELO_CELL_AND_PORT`    | Tie low cells used in Yosys synthesis to replace a logical 0 in the Netlist.               |
| `MIN_BUF_CELL_AND_PORTS` | Used to insert a buffer cell to pass through wires. Used in synthesis.                     |
| `ABC_CLOCK_PERIOD_IN_PS` | Clock period to be used by STA during synthesis. Default value read from `constraint.sdc`. |
| `ABC_DRIVER_CELL`        | Default driver cell used during ABC synthesis.                                             |
| `ABC_LOAD_IN_FF`         | During synthesis set_load value used.                                                      |

### Floorplan

| Variable              | Description                                                                                                                                                                      |
|-----------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `PLACE_SITE`          | Placement site for core cells defined in the `.lef`.                                                                                                                             |

Reviewer: ```suggestion
| `PLACE_SITE`          | Placement site for core cells defined in the technology LEF file.                                                                                                                |

```

| `TAPCELL_TCL`         | Path to Endcap and Welltie cells file.                                                                                                                                           |
| `MACRO_PLACE_HALO`    | horizontal/vertical halo around macros (microns).                                                                                                                                |
| `MACRO_PLACE_CHANNEL` | horizontal/vertical channel width between macros (microns).                                                                                                                      |
| `PDN_CFG`             | File path which has a set of power grid policies used by `pdn` to be applied to the design, such as layers to use, stripe width and spacing to generate the actual metal straps. |
| `MAKE_TRACKS`         | Tcl file that defines add routing tracks to a floorplan.                                                                                                                         |

> | `IO_PIN_MARGIN`       | Maximum number of I/O pins placed outside floorplan boundary. Integer value.                                                                                                     |

Reviewer: This variable is deprecated and should be removed.

`./flow/designs/<PLATFORM>/<DESIGN_NAME>/config.mk`

| Variable        | Description                                                                                                                                                          |
|-----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `PLATFORM`      | Specifies process design kit or technology node to be used.                                                                                                          |
| `DESIGN_NAME`   | The name of the top-level module of the design.                                                                                                                      |
| `VERILOG_FILES` | The path to the design Verilog files.                                                                                                                                |
| `SDC_FILE`      | The path to design constraint (SDC) file.                                                                                                                            |

> | `CLOCK_PORT`    | The name of the designs clock port used in Static Timing Analysis. If your design does not have a clock port defined as an empty variable `export CLOCK_PORT= ""`. |

Reviewer: Deprecated / unused.

> | `CLOCK_PERIOD`  | The clock period for the design to meet its goal.                                                                                                                    |

Reviewer: Deprecated / unused

### Optional Variables

These are optional variables that may be over-ridden/appended with
default value from the platform `config.mk` by defining in the design
configuration file.

#### Synthesis

| Variable                 | Description                                                                                        |
|--------------------------|----------------------------------------------------------------------------------------------------|
| `ADDITIONAL_LEFS`        | Hardened macro LEF view files listed here.                                                         |
| `ADDITIONAL_LIBS`        | Hardened macro library files listed here.                                                          |
| `ADDITIONAL_GDS`         | Hardened macro GDS files listed here.                                                              |
| `VERILOG_INCLUDE_DIRS`   | Specifies the include directories for the Verilog input files.                                     |

> | `VERILOG_FILES_BLACKBOX` | Hardened macro Verilog file list to be added here.                                                 |

Reviewer: This isn't an official flow variable. If it was it's deprecated and not used anywhere.

> | `CORNER`                 | PVT corner library selection fast/typical/worst.                                                   |

Reviewer: Corner doesn't have to be fast/typical/worst.
```suggestion
| `CORNER`                 | PVT corner library selection.                                                                      |

```

> | `DFF_LIB_FILE`           | Define this variable if the design has asynchronous clocks.                                        |

Reviewer: This is just a hack for ASAP7. It isn't used for any other platform.

> | `DESIGN_NICKNAME`        | If the design has a hierarchy, to select a specific module as top design this variable is defined. |

Reviewer: `DESIGN_NICKNAME` is purely for convenience. It doesn't affect synthesis. It just changes the directory name that ORFS outputs to be `DESIGN_NICKNAME` instead of `DESIGN_NAME` in case `DESIGN_NAME` is unwieldy or conflicts with a difference design.

| `ABC_AREA`               | Strategies for Yosys ABC synthesis: Area/Speed. Default ABC_SPEED.                                 |
| `PWR_NETS_VOLTAGES`      | Used for IR Drop calculation.                                                                      |
| `GND_NETS_VOLTAGES`      | Used for IR Drop calculation.                                                                      |

#### Floorplan

| Variable              | Description                                                                                                                                                                                                                                |
|-----------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `CORE_UTILIZATION`    | The core utilization percentage (0-100). Overrides DIE_AREA & CORE_AREA                                                                                                                                                                    |

Reviewer: ```suggestion
| `CORE_UTILIZATION`    | The core utilization percentage (0-100). Overrides DIE_AREA & CORE_AREA.                                                                                                                                                                   |

```

> | `CORE_ASPECT_RATIO`   | The core aspect ratio (height / width). If CORE_UTILIZATION defined, it can be defined else ignore it.                                                                                                                                     |

Reviewer: ```suggestion
| `CORE_ASPECT_RATIO`   | The core aspect ratio (height / width). This value is ignored if `CORE_UTILIZATION` is undefined.                                                                                                                               |

```

> | `CORE_MARGIN`         | The core margin, in multiples of site heights, from the bottom boundary. If CORE_UTILIZATION defined, it can be defined else ignore it.                                                                                                    |

Reviewer: ```suggestion
| `CORE_MARGIN`         | The margin between the core area and die area, in multiples of SITE heights. The margin is applied to each side. This variable is ignored if `CORE_UTILIZATION` is undefined.                                                         |

```

> | `DIE_AREA`            | Specify die area to be used in floorplanning. Specified as a 4-corner rectangle.Ignored if CORE_UTILIZATION is set. Unit microns.                                                                                                          |

Reviewer: ```suggestion
| `DIE_AREA`            | The die area specified as a list of lower-left and upper-right corners in microns (X1 Y1 X2 Y2). This variable is ignored if `CORE_UTILIZATION` and `CORE_ASPECT_RATIO` are defined.                           |

```

> | `CORE_AREA`           | Specify core area to be used in floorplanning. Specified as a 4-corner rectangle. The core area can be specified in absolute terms and is mutually exclusive from `CORE_UTILIZATION` and `CORE_ASPECT_RATIO` specifications. Unit microns. |

Reviewer: ```suggestion
| `CORE_AREA`           | The core area specified as a list of lower-left and upper-right corners in microns (X1 Y1 X2 Y2). This variable is ignored if `CORE_UTILIZATION` and `CORE_ASPECT_RATIO` are defined.                                                                 |

```
